libraries/WiFi/src/WiFiUdp.cpp

```diff
@@ -207,22 +291,44 @@ size_t WiFiUDP::write(const uint8_t *buffer, size_t size){
 int WiFiUDP::parsePacket(){
   if(rx_buffer)
     return 0;
-  struct sockaddr_in si_other;
-  int slen = sizeof(si_other) , len;
+  struct sockaddr_storage si_other_storage;   // enough storage for v4 and v6
+  socklen_t slen = sizeof(sockaddr_storage);
+  int len;
   char *buf = (char *)malloc(1460);
   if(!buf) {
     return 0;
   }
-  if ((len = recvfrom(udp_server, buf, 1460, MSG_DONTWAIT, (struct sockaddr *) &si_other, (socklen_t *)&slen)) == -1){
+  if ((len = recvfrom(udp_server, buf, 1460, MSG_DONTWAIT, (struct sockaddr *) &si_other_storage, (socklen_t *)&slen)) == -1){
     free(buf);
     if(errno == EWOULDBLOCK){
       return 0;
     }
     log_e("could not receive data: %d", errno);
     return 0;
   }
-  remote_ip = IPAddress(si_other.sin_addr.s_addr);
-  remote_port = ntohs(si_other.sin_port);
+  if (si_other_storage.ss_family == AF_INET) {
+    struct sockaddr_in &si_other = (sockaddr_in&) si_other_storage;
+    remote_ip = IPAddress(si_other.sin_addr.s_addr);
+    remote_port = ntohs(si_other.sin_port);
+  }
+#if LWIP_IPV6 
+  else if (si_other_storage.ss_family == AF_INET6) {
+    struct sockaddr_in6 &si_other = (sockaddr_in6&) si_other_storage;
+    remote_ip = IPAddress(IPv6, (uint8_t*)&si_other.sin6_addr, si_other.sin6_scope_id);   // force IPv6
+    ip_addr_t addr;
+    remote_ip.to_ip_addr_t(&addr);
+    /* Dual-stack: Unmap IPv4 mapped IPv6 addresses */
+    if (remote_ip.type() == IPv6 && ip6_addr_isipv4mappedipv6(ip_2_ip6(&addr))) {
+      unmap_ipv4_mapped_ipv6(ip_2_ip4(&addr), ip_2_ip6(&addr));
+      IP_SET_TYPE_VAL(addr, IPADDR_TYPE_V4);
```

Comment:
Sorry there is one last problem here. `IP_SET_TYPE_VAL(addr, IPADDR_TYPE_V4);` changes `addr` but not `remote_ip` so this code does nothing.

So I propose instead:
`remote_ip = IPAddress(addr.u_addr.ip4.addr);`

Or if you prefer:
`remote_ip.from_ip_addr_t(addr);`

Reply:
we do have the constructor now :)

Reply:
done

Reply:
Awesome, now it's all good for me.